platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java

```diff
@@ -236,6 +246,99 @@ public String toString() {
+    }
+
+    /**
+     * Handle a classpath that can potentially be the output of the instrumenting artifact transform. The rules for conversion are:
+     * <ol>
+     *     <li>The {@link TransformedClassPath} is returned as is, because it already bears the instrumentation mappings.</li>
+     *     <li>The classpath that only contains original JARs is returned as is, because there's no instrumentation mappings to apply.</li>
+     *     <li>The classpath that conforms to {@link #fromInstrumentingArtifactTransformOutput(ClassPath)} requirements is converted to {@link TransformedClassPath} accordingly.</li>
+     * </ol>
+     *
+     * @param classPath the classpath to process
+     * @return the classpath that carries the instrumentation mappings if needed
+     */
+    public static ClassPath handleInstrumentingArtifactTransform(ClassPath classPath) {
+        if (classPath.isEmpty() || classPath instanceof TransformedClassPath) {
+            return classPath;
+        }
+
+        int resultSize = computeResultSizeOfInstrumentingArtifactTransformOutput(classPath);
+        if (resultSize == classPath.getAsFiles().size()) {
+            // There is no instrumented JARs in the list, so the classpath can be used without transformation.
+            return classPath;
+        }
+
+        return fromInstrumentingArtifactTransformOutputWithExpectedSize(classPath, resultSize);
+    }
+
+    /**
+     * Each instrumented JAR has a corresponding original entry, but not necessarily vice versa, so the number of originals is the size of the result.
+     */
+    private static int computeResultSizeOfInstrumentingArtifactTransformOutput(ClassPath classPath) {
+        int resultSize = 0;
+        for (int i = 0; i < classPath.getAsFiles().size(); ++i) {
+            File inputFile = classPath.getAsFiles().get(i);
+            if (isInstrumentedMarkerFile(inputFile)) {
+                // Marker file will be followed by instrumented jar and original jar.
+                i += 2;
+            }
+            ++resultSize;
+        }
+        return resultSize;
+    }
+
+    private static TransformedClassPath fromInstrumentingArtifactTransformOutputWithExpectedSize(ClassPath classPath, int resultSize) {
+        List<File> inputFiles = classPath.getAsFiles();
+
+        Builder result = builderWithExactSize(resultSize);
+        for (int i = 0; i < inputFiles.size(); ++i) {
+            File inputFile = inputFiles.get(i);
+            if (isInstrumentedMarkerFile(inputFile)) {
+                checkArgument(i + 2 < inputFiles.size(), "Missing the instrumented or original JAR for classpath %s", inputFiles);
+                File instrumentedJar = inputFiles.get(i + 1);
+                File originalJar = inputFiles.get(i + 2);
+                checkArgument(areInstrumentedAndOriginalJarValid(instrumentedJar, originalJar), "Instrumented JAR doesn't match original JAR %s", originalJar.getName());
```

Comment:
:nail_care: we should probably also add the instrumented jar name too. And I think printing full paths would be better, because e.g. parent dirs participate in the check too.
```suggestion
                checkArgument(areInstrumentedAndOriginalJarValid(instrumentedJar, originalJar), "Instrumented JAR %s doesn't match original JAR %s", instrumentedJar.getAbsolutePath(), originalJar.getAbsolutePath());
```

Reply:
Makes sense